File: brains/datafile.py
import random

class DataFile(object):
    """Reads csv files containing class, feature lists. Streams to 
    save memory."""

    SEP = ","

    def __init__(s, path):
        s.file = open(path, "r")
        s.features = None
        s.count = 0
        for line in s.file:
            if line.find("#") == 0:
                continue
            parts = line.strip().split(DataFile.SEP)
            if s.features == None:
                s.features = parts[1:]
                break

    def points(s):
        for line in s.file:
            parts = line.strip().split(DataFile.SEP)
            klass = parts[0]
            data = zip(s.features, parts)
            s.count += 1
            yield(klass, dict(data))

    def split(s, fraction):
        allp = [p for p in s.points()]
        random.shuffle(allp)
        split = len(allp) / fraction
        return (allp[:split], allp[split:])
        

File: lib/datafile.rb
# datafile loading class
class DataFile

  include Enumerable

  attr_reader :size, :names, :samples
  SEP = /\s*,\s*/

  def initialize(path)
    @file = open(path)
    while line = @file.readline
      break unless line =~ /^\s*#/ # skip comments
    end
    klass, *@names = line.strip.split(SEP)
    @names.collect! { |i| i.to_sym }
    @size = 0
  end

  def each
    @file.each_line do |f|
      @size += 1
      klass, *parts = f.strip.split(SEP)
      next unless parts.size == @names.size
      parts = Hash[@names.zip(parts)]
      yield [klass, parts]
    end
  end

  def split(fraction, shuffle = false)
    all = to_a
    all = all.shuffle if shuffle
    index = all.size / fraction
    [all[0..index-1], all[index..-1]]
  end

end
